Improve the application structure documentation (#478)
encoredev/encore: docs/develop/app-structure.md

---

Encore uses a monorepo design and it's best to use one Encore app for your entire backend application.
This lets Encore build an application model that spans your entire app, necessary to get the most value out of many
features like [distributed tracing](/docs/observability/tracing) and [Encore Flow](/docs/develop/encore-flow).

If you have a large application, see advice on how to [structure an app with several systems](/docs/develop/app-structure#large-applications-with-several-systems). 

It's easy to integrate Encore applications with any pre-existing systems you might have. To make this frictionless,
Encore comes with built-in tools like [client generation](/docs/develop/client-generation).

## Monoliths and small microservices applications

> Small architectures like monoliths, or apps with a small number of services, can be structure as a list of Encore
> services. Services are the base building blocks of your application, they separate your application in isolated units
> with their own endpoints and database (if using one). In a microservices application, an Encore service represents a
> single microservice. While services are isolated and each have their own databases by default,
> [databases may be shared between services](/docs/how-to/share-db-between-services).

Reviewer: ```suggestion
Small architectures like monoliths, or apps with a small number of services, can be structured using one or many Encore
services. Services are the base building blocks of your application and divide it into isolated units
with their own endpoints and database(s). In a microservices application, an Encore service represents a
single microservice. While services are isolated and have their own databases by default,
[databases can also be shared between services](/docs/how-to/share-db-between-services).
```

Reviewer: I thought this part was a bit confusing using terminology like "list".
Also tweaked it in trying to keep it brief.

with their own endpoints and database (if using one). In a microservices application, an Encore service represents a
single microservice. While services are isolated and each have their own databases by default,
[databases may be shared between services](/docs/how-to/share-db-between-services).

> Simply create a Go package for each service (also known as service packages), [define APIs](/docs/develop/services-and-apis) and business logic in Go files, and add database migrations to define the structure of any database(s). (See more about [SQL databases](/docs/develop/databases).)

Author: I merged the text from main here and remove any words like "simply" in this tutorial since it's usually better to avoid any word that imply how easy things should be for users in documentation.

Reviewer: Great point!

    └── world.go                     // world service code
```

Encore is not opinionated about whether you use a monolith or multiple services. However, it does solve most of the
traditional drawbacks that come with building microservices.

> ## Define service code with sub-packages

Reviewer: I think "service code" is too general to be clearly understood?

> Within a service, it's possible to have multiple sub-packages. This is a good way to define components, helper
> functions, or other code for your functions, should you wish to do that. You can create as many subpackages in any
> kind of nested structure to build your service in any way you want.

Reviewer: ```suggestion
Within a service, it's possible to have multiple sub-packages. This is a good way to define components, helper
functions, or other code for your functions, should you wish to do that. You can create as many sub-packages, in any kind of nested structure within your service, as you want.
```

Reviewer: This is a great addition, just trying to clarify the last sentence a bit.

To create sub-packages, you create sub-directories within a service package. Subpackages are itnernal code for services,
they are not service packages. Only service packages can define API endpoints, and sub-packages within services cannot
themselves define APIs.
You can however define an API in a service package that calls a function within a sub-package.

> For example, rather than define the entire logic for an endpoint in that endpoint's function, you can call functions

Author: I added this example here since I felt this section was a bit confusing? I think the goal here was to explain that you can have more subpackages than just the service package, hopefully it's clear. Should we maybe also mention the possibility to have internal packages?

> To create sub-packages, you create sub-directories within a service package. Subpackages are itnernal code for services,
> they are not service packages. Only service packages can define API endpoints, and sub-packages within services cannot
> themselves define APIs.
> You can however define an API in a service package that calls a function within a sub-package.
> 
> For example, rather than define the entire logic for an endpoint in that endpoint's function, you can call functions

Reviewer: ```suggestion
To create sub-packages, you create sub-directories within a service package. Sub-packages are internal to services,
they are not themselves service packages. This means sub-packages within services cannot
themselves define APIs.
You can however define an API in a service package that calls a function within a sub-package.

For example, rather than define the entire logic for an endpoint in that endpoint's function, you can call functions
```

Reviewer: Fixed some spelling and more consistent use of `sub-packages` (vs. `subpackages`)

        └── user.go             // user service code
```

> The only refactoring needed to separate a service application into systems is to move services into their respective
> subfolders. This greatly simplifies the application around the specific concerns of each system. For encore, what
> matters are the packages containing services. The division in systems or subsystems will not change the endpoint and
> architecture of your application

Reviewer: ```suggestion
The only refactoring needed to divide an existing Encore application into systems is to move services into their respective
subfolders. This is a simple way to separate the specific concerns of each system. What
matters for Encore are the packages containing services, and the division in systems or subsystems will not change the endpoints or
architecture of your application.
```

Reviewer: "Service application" seemed a bit vague. So tweaked it to be a bit more use case centric, i.e. go from encore app without systems to one with systems.
